### PR TITLE
Model#deletePerson(...): Change parameter from ReadOnlyPerson to Index #465

### LINKED ISSUE #465: Change Model.deletePerson(ReadOnlyPerson) to Model.deletePerson(Index)
Command has no need to know which `ReadOnlyPerson` is being involved in the deletion process. Furthermore, we should standardise with other API in `Model` such as #updatePerson which takes in an index as well.

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #465

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/1] [Update Model\#deletePerson\(ReadOnlyPerson\)](https://github.com/se-edu/addressbook-level4/pull/466/commits/8923a504781796a7aae14b10799789cd04d58143)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/dfe3b25eb274119f2d0605fa266877597a3341c6...8923a504781796a7aae14b10799789cd04d58143))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/466/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 1]
```diff
@@ -368,7 +368,7 @@ public void execute_delete_removesCorrectPerson() throws Exception {
         List<Person> threePersons = helper.generatePersonList(3);
 
         Model expectedModel = new ModelManager(helper.generateAddressBook(threePersons), new UserPrefs());
-        expectedModel.deletePerson(threePersons.get(1));
+        expectedModel.deletePerson(Index.fromOneBased(2));
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: Can use `INDEX_SECOND_PERSON`?

**Comment by yamgent** [comment]
Also the commit message title / PR title doesn't seem to give an indication that we are moving to `Index`.

**Comment by yamgent** [comment]
Suggested title: `Model#deletePerson(...): Change parameter from ReadOnlyPerson to Index`

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/1] [Model\#deletePerson\(\.\.\.\): Change parameter from ReadOnlyPerson to Index](https://github.com/se-edu/addressbook-level4/pull/466/commits/420d8fff327cb265f6a38b7902ae540f632eeba0)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/dfe3b25eb274119f2d0605fa266877597a3341c6...420d8fff327cb265f6a38b7902ae540f632eeba0)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/d1662c35e051245bb0c8df9022ebd3360a3a39e0~1...d1662c35e051245bb0c8df9022ebd3360a3a39e0))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/466/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/1] [Model\#deletePerson\(\.\.\.\): Change parameter from ReadOnlyPerson to Index](https://github.com/se-edu/addressbook-level4/pull/466/commits/6bcc7a5a2b4e0636797c0150ea5bcd72319f66b1)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/dfe3b25eb274119f2d0605fa266877597a3341c6...6bcc7a5a2b4e0636797c0150ea5bcd72319f66b1)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/17b1de475c83211b34adba89aab048f6efa73327~1...17b1de475c83211b34adba89aab048f6efa73327))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/466/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by damithc** [comment]
Now that I think about this, may be the Model interface should not use index at all, as the index is related to internal data structure?

File: src/main/java/seedu/address/model/Model.java [hunk 2]
```diff
@@ -18,8 +18,8 @@
     /** Returns the AddressBook */
     ReadOnlyAddressBook getAddressBook();
 
-    /** Deletes the given person. */
-    void deletePerson(ReadOnlyPerson target) throws UniquePersonList.PersonNotFoundException;
+    /** Deletes the person at {@code index}. */
```
Comment (MightyCupcakes) [inline src/main/java/seedu/address/model/Model.java]: Index of what? If I were not familiar of the code, this javadoc is completely ambiguous.

Also what happens if i did this:

```java
model.deletePerson(new Index(Integer.MAX_VALUE));
```

is it covered by the tests? Ignore if this case is covered; I havent had the time to check it

**Comment by yamgent** [comment]
@Zhiyuan-Amos The argument here is that perhaps we should change `updatePerson(int, ReadOnlyPerson)` instead. (i.e. it should be `Model#updatePerson(ReadOnlyPerson oldPerson, ReadOnlyPerson newPerson)`).

**Comment by damithc** [comment]
At first I thought the `index` in `delete(index)` refers to the index in the main Person list, in which case I felt it is exposing an internal detail via the API. But now I realize that the index refers to the filtered list. Even in that case, using it in the API makes the method less cohesive because it translates to `delete the person that is shown as 3rd person in the filtered list` whereas it is more cohesive to say `delete this guy` or `delete Adam Boon` (provided we don't allow duplicate names). Imagine in future we decide to move the filtered list out of the model. That should not force us to modify the `delete()` method because deleting a person has nothing to do with filtering as far as the Model is concerned. Coming from another angle, model has two parts: main list and the filtered list. `delete(index)` introduces an indirect coupling from the main list to the filtered list which we can do without.

**Pull request closed without merging**